### PR TITLE
Fix checkerboarding 

### LINKED ISSUE #276: Fix checkerboarding in convolutional decoder
### Problem / Goal

Currently the convolutional decoder shows checkerboarding artifacts. [This page](https://distill.pub/2016/deconv-checkerboard/) gives a good overview of why this happens and how to fix it.

### What does good look like?

- [ ] No checkerboarding artifacts in the convolutional decoder

### PR DESCRIPTION
**Pull request by jemrobinson**
Checkerboarding was caused by convolutional overlap in ConvBlockUpsample as discussed [here](https://distill.pub/2016/deconv-checkerboard/).

Simply forcing the kernel size to be a multiple of the stride was insufficient. The eventual solution was to implement [subpixel upsampling](https://arxiv.org/abs/1609.05158) with [ICNR initialisation](https://arxiv.org/abs/1707.02937).

<details>
<summary>Config file overrides</summary>

```
_target_: icenet_mp.models.EncodeProcessDecode

name: 276-checkerboard

encoders:
  latent_space: [108, 108]
  era5:
    _target_: icenet_mp.models.encoders.CNNEncoder
    n_layers: 2
  sic-icenet:
    _target_: icenet_mp.models.encoders.CNNEncoder
    n_layers: 2

processor:
  _target_: icenet_mp.models.processors.NullProcessor

decoder:
  _target_: icenet_mp.models.decoders.CNNDecoder
  n_layers: 2
  bounded: false

predict:
  n_forecast_steps: 1
  n_history_steps: 1

train:
  optimizer:
    lr: 0.01
  trainer:
    max_epochs: 10
```
</details>

Closes #276

## Before: b01e4fec

main@b01e4fec

See the noticeable aliasing pattern especially near the Weddell Sea.

## Commit: cf2b845

main@f948dae

Just changing the kernel size isn't enough to prevent checkerboarding.

## Commit: 07e0d726

main@07e0d726

Ignoring the subzero predictions for the sea (showing as white) the Weddell Sea prediction is much improved but still blurry and underpredicting.

## Commit: 13d3cd1

13d3cd1

Switching to subpixel upsampling gives the best performance.

### CODE REVIEW

File: icenet_mp/models/common/conv_block_upsample.py [hunk 1]
```diff
@@ -6,9 +6,9 @@
 
 
 class ConvBlockUpsample(nn.Module):
-    """Convolutional block that doubles the resolution.
+    """Convolutional block that doubles each spatial dimension and halves the number of channels.
 
-    (ConvTranspose2d → Normalization → Activation) → (ConvTranspose2d → Normalization → Activation)
+    (ConvTranspose2d > Normalization > Activation) > (ConvTranspose2d > Normalization > Activation)
 
     This is the reverse of ConvBlockDownsample.
```
[the remark this answers is not in the document]
Comment (jemrobinson) [reply icenet_mp/models/common/conv_block_upsample.py]: Changed to 276

File: icenet_mp/models/common/conv_block_upsample.py [hunk 2]
```diff
@@ -25,7 +25,7 @@ def __init__(
 
         Args:
             activation: the activation function to use.
-            kernel_size: the size of the convolutional kernel (odd numbers are preferable!).
+            kernel_size: the size of the convolutional kernel.
```
[the remark this answers is not in the document]
Comment (jemrobinson) [reply icenet_mp/models/common/conv_block_upsample.py:35]: Done in 872fa85a

File: icenet_mp/models/common/conv_norm_act_upsample.py [hunk 3]
```diff
@@ -34,8 +34,9 @@ def __init__(
         """
         super().__init__()
 
+        out_channels = in_channels // 2 if out_channels is None else out_channels
         self.block = nn.Sequential(
```
[the remark this answers is not in the document]
Comment (jemrobinson) [reply icenet_mp/models/common/conv_norm_act_upsample.py:46]: Done in 872fa85a

File: icenet_mp/models/common/conv_block_upsample_naive.py [hunk 4]
```diff
@@ -34,8 +34,9 @@ def __init__(
         """
         super().__init__()
 
+        out_channels = in_channels // 2 if out_channels is None else out_channels
         self.block = nn.Sequential(
-            nn.Upsample(scale_factor=2, mode="nearest"),
+            nn.Upsample(scale_factor=2, mode="bilinear"),
```
[the remark this answers is not in the document]
Comment (jemrobinson) [reply icenet_mp/models/common/conv_block_upsample_naive.py]: Reverted in 92962870

File: icenet_mp/models/common/conv_block_upsample.py [hunk 5]
```diff
@@ -37,30 +37,31 @@ def __init__(
         n_output_channels = (
             n_input_channels // 2 if n_output_channels is None else n_output_channels
         )
-        padding = (kernel_size - 1) // 2
-        output_padding = kernel_size % 2
+        kernel_size_odd = kernel_size if kernel_size % 2 else kernel_size + 1
+        kernel_size_even = kernel_size + 1 if kernel_size % 2 else kernel_size
```
[the remark this answers is not in the document]
Comment (jemrobinson) [reply icenet_mp/models/common/conv_block_upsample.py]: Done in 68a76cb4

File: icenet_mp/models/common/conv_block_upsample.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jemrobinson) [reply icenet_mp/models/common/conv_block_upsample.py]: Done in 1c87cbc2

File: icenet_mp/models/common/conv_block_upsample_naive.py [hunk 6]
```diff
@@ -24,16 +24,17 @@ def __init__(
...
 
         Args:
-            in_channels (int): Input channel size.
-            out_channels (int): Output channel size.
-            kernel_size (int): Kernel size for the convolution after upsampling.
-            norm_type (str): Type of normalization ("groupnorm", "batchnorm", or "none").
-            activation (str): Name of activation function.
-            dropout_rate (float): Dropout probability for ConvNormAct.
+            in_channels: Number of input channels.
+            out_channels: Number of output channels (if None, half of n_input_channels).
```
[the remark this answers is not in the document]
Comment (jemrobinson) [reply icenet_mp/models/common/conv_block_upsample_naive.py]: Done in 1c87cbc2

File: icenet_mp/models/common/conv_norm_act_upsample.py [hunk 7]
```diff
@@ -4,36 +4,37 @@
...
-    Note that using ConvTranspose2d as implemented in ConvBlockUpsample, is generally
-    preferred (see e.g. https://discuss.pytorch.org/t/upsample-conv2d-vs-convtranspose2d/138081)
+    Prefer ConvBlockUpsample for most use cases (see https://discuss.pytorch.org/t/upsample-conv2d-vs-convtranspose2d/138081).
     """
 
     def __init__(
         self,
         in_channels: int,
-        out_channels: int,
+        *,
```
[the remark this answers is not in the document]
Comment (jemrobinson) [reply icenet_mp/models/common/conv_norm_act_upsample.py:13]: Done in cf2b845d

File: icenet_mp/models/common/conv_block_upsample_naive.py [hunk 8]
```diff
@@ -4,36 +4,37 @@
...
 class ConvBlockUpsampleNaive(nn.Module):
-    """Convolutional block that doubles the resolution.
+    """Convolutional block that reduces channels and doubles each spatial dimension.
 
-    Upsample → ConvNormAct
+    Upsample > (Conv2d > Normalization > Activation)
 
-    Note that using ConvTranspose2d as implemented in ConvBlockUpsample, is generally
-    preferred (see e.g. https://discuss.pytorch.org/t/upsample-conv2d-vs-convtranspose2d/138081)
+    Prefer ConvBlockUpsample for most use cases (see https://discuss.pytorch.org/t/upsample-conv2d-vs-convtranspose2d/138081).
```
[the remark this answers is not in the document]
Comment (jemrobinson) [reply icenet_mp/models/common/conv_block_upsample_naive.py]: Done in 6e79b3e7

**Review by IFenton — APPROVED** [review]
Can you set these parameters in the yaml file? 

But LGTM

**Comment by jemrobinson** [comment]
@IFenton : just re-checking this after recent changes to main. Which parameters would you like to be able to set in the YAML file?

File: icenet_mp/models/common/normalisations.py [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+def normalisation_from_name(norm_type: str, out_channels: int) -> nn.Module:
+    """Create a normalization layer based on the specified type."""
+    # Batch normalisation
+    if norm_type.lower() == "batchnorm":
+        return nn.BatchNorm2d(out_channels)
+
+    # Group normalisation
+    if norm_type.lower() == "groupnorm":
+        # Determine the highest integer less than 8 that divides `channels`
+        num_groups = max(num for num in range(1, 8) if out_channels % num == 0)
```
[the remark this answers is not in the document]
Comment (jemrobinson) [reply icenet_mp/models/common/normalisations.py]: Done in fa4bf662

File: icenet_mp/models/diffusion/unet_diffusion.py [hunk 10]
```diff
@@ -75,7 +75,7 @@ def __init__(  # noqa: PLR0913
         # Encoder
         self.conv1 = CommonConvBlock(
             in_channels=input_channels
-            + output_channels,  # input_channels (n_input_days * n_input_channels), output_channels (n_output_classes * n_forecast_days)
+            + output_channels,  # input_channels (n_input_days * input_channels), output_channels (n_output_classes * n_forecast_days)
```
[the remark this answers is not in the document]
Comment (jemrobinson) [reply icenet_mp/models/diffusion/unet_diffusion.py]: Done in 3cd0058f

File: icenet_mp/models/common/conv_block_downsample.py [hunk 11]
```diff
@@ -6,46 +6,55 @@
 
 
 class ConvBlockDownsample(nn.Module):
-    """Convolutional block that halves the resolution and doubles the number of channels.
+    """Convolutional block that halves spatial dimensions.
 
-    This is the reverse of ConvBlockUpsample.
+    ConvNormAct > ConvNormAct
+
+    If out_channels is not specified than this will double the number of input channels.
```
[the remark this answers is not in the document]
Comment (jemrobinson) [reply icenet_mp/models/common/conv_block_downsample.py]: Done in c8e00f3a

File: icenet_mp/models/common/conv_block_upsample.py [hunk 12]
```diff
@@ -3,67 +3,65 @@
...
 
 class ConvBlockUpsample(nn.Module):
-    """Convolutional block that doubles the resolution.
+    """Convolutional block that doubles spatial dimensions.
 
-    (ConvTranspose2d → Normalization → Activation) → (ConvTranspose2d → Normalization → Activation)
+    (WeightedUpsample > Norm > Act) > ConvNormAct > ConvNormAct
 
-    This is the reverse of ConvBlockDownsample.
+    If out_channels is not specified than this will halve the number of input channels.
```
[the remark this answers is not in the document]
Comment (jemrobinson) [reply icenet_mp/models/common/conv_block_upsample.py]: Done in c8e00f3a

File: icenet_mp/models/common/normalisations.py [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+
+def normalisation_from_name(norm_type: str, out_channels: int) -> nn.Module:
+    """Create a normalization layer based on the specified type."""
+    # Batch normalisation
+    if norm_type.lower() == "batchnorm":
+        return nn.BatchNorm2d(out_channels)
+
+    # Group normalisation
+    if norm_type.lower() == "groupnorm":
+        # Determine the highest integer less than 8 that divides `channels`
```
[the remark this answers is not in the document]
Comment (jemrobinson) [reply icenet_mp/models/common/normalisations.py]: Done in c8e00f3a
